rfcs/90-parse-pairs-from-string.md

```diff
@@ -0,0 +1,106 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-01
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Parse Pairs from String
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose the following mechanism for parsing pairs:
+
+### Pair Registry
+
+We first register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]reflect.Type
+
+func RegisterServicePairMap(ty string, m PairMap) {}
+```
+
+From services side, we can generate following code:
+
+```go
+func servicePairs() PairMap {}
+
+func init() {
+	// ...
+	pairs.RegisterServicePairMap("<type>", servicePairs())
+}
+```
+
+The pair information maps are also generated.
+
+### Parse According to Type
+
+For each type, we can have a parse function.
+
+```go 
+type parseFunc func(string) (interface{}, error)
```

Comment:
`parseFunc func(string) (interface{}, error)` not looks good to me.

`definitions` already has all information to parse the string into strongly typed pairs.

Reply:
I didn't get it. How does `definitions` parse a string into a value?

Reply:
Since from pair name and type in toml we can get a `WithXxx(v T)`, we should be able to get `T` from `string`. But how to parse `string` into the `v` is not decided, I think. (so we will need a `parseFunc`?)

Reply:
BTW, things like `typeMap` are private in `definitions`. How can we use them?

Reply:
> But how to parse string into the v is not decided, I think. 

It's what we need to specify in this GSP, for example:

From 

```
s3://<credential>@<endpoint>/workdir?key=value&default_storage_pairs.write=[key=value&key2=value]
```

To s3 and related pairs.

> BTW, things like typeMap are private in definitions.

We don't get them from `typeMap`.

Our code should be generated by `definitions` and we can get that information from specs and `service.toml`.

`service.toml` defined all pairs that this service supported.

Reply:
Let me make things more clear.
In package `pairs`, we add suport to `parse(name string, value string) Pair`  (what puclic API is doesn't matter, internally there should be one like this), so we should 
- know `name` is a pair and its type `T`.  
- parse string into `T`

So in `pairs` we use a `PairMap` to register pairs. The specific registration code is generated by `definition`. `definition` knows everything and can generate type information both as string and as `T`, but what should be contained by `PairMap`? 

After code generation, `definitions` is useless. Then to parse string into `T` in `pairs`, we should know `T` and its parse function. If the registry `PairMap` stores string, how can we know what `T` is? And what's wrong that the parse function is `parseFunc func(string) (interface{}, error)`?

Reply:
TLDR: `definitions` only generate registration code which sets `PairMap`. Any other logic, include parsing `string` into `T` is written in `pairs`.

If this is not true, please tell me

Reply:
Oh I think we can
```go
typ := pairMap[name]
switch typ {
case "int": 
    pair.Value=parseInt(value)
case "bool":
    pair.Value=parseBool(value)
...
}
```
 
I used `reflect.Type` and `parseFunc func(string) (interface{}, error)` just to save the work here... I hoped to be more elegant and extensible, but actually this simple way is better, I think

Reply:
> Oh I think we can
> 
> ```go
> typ := pairMap[name]
> switch typ {
> case "int": 
>     pair.Value=parseInt(value)
> case "bool":
>     pair.Value=parseBool(value)
> ...
> }
> ```

Nice catch.

We don't need to support `parse(name string, value string) Pair`, we only need to parse the value and always transform them together instead of one by one.

Reply:
OK, I'll update the proposal and demo later, although I'm still not 100% sure of what you mean. Anyway, let's delay the discussion after the update.